Implement DeadCom Layer 2, migrate docs

adam-dej opened the pull request.

adam-dej committed:
Add .gitignore

adam-dej committed:
Import yahdlc library, remove global state

adam-dej committed:
Add deadcom header

adam-dej committed:
Update licence notices

adam-dej committed:
Add Readme and restructure repo

adam-dej committed:
Add Unity dependency

adam-dej committed:
Port unit tests from yahdlc project

adam-dej committed:
Add unit tests for custom yahdlc features

adam-dej committed:
Fix U-frame decoding details in yahdlc

adam-dej committed:
Add FFF dependency, declare sync primitives and add init tests

adam-dej committed:
[dcl2] Implement dcConnect

adam-dej committed:
[dcl2] Implement dcDisconnect

adam-dej committed:
[dcl2] Fix yahdlc incorrect I frame generation and decoding problems

adam-dej committed:
[dcl2] Implement dcSendMessage

adam-dej committed:
[dcl2] Implement retrieving received messages

adam-dej committed:
[dcl2] Add extractionComplete flag

adam-dej committed:
[dcl2] Change yahdlc indentation to 4 spaces

adam-dej committed:
[dcl2] Yahdlc can now compute frame lengths before their creation

adam-dej committed:
[dcl2] Run tests with valgrind

adam-dej committed:
[dcl2] dcProcessData skelet and processing in Disconnected mode

adam-dej committed:
[dcl2] Input data processing in Connecting mode

adam-dej committed:
[dcl2] Fix incorrect yahdlc_get_data mock

adam-dej committed:
[dcl2] Remove DISCONNECTED frame tx tests, this frame will be deprecated

adam-dej committed:
[dcl2] Implement Data frame processing

adam-dej committed:
Add Gitlab-CI config for automated testing

adam-dej committed:
Disable trampoline warnings when compiling tests

adam-dej committed:
Migrate documentation for protocols to this repo

adam-dej committed:
Add ci config for test-building docs

adam-dej committed:
[dcl2] Remove DISCONNECTED frame

adam-dej committed:
[dcl2] Remove code duplicity from tests

adam-dej committed:
[dcl2] Fix backwards assertions

adam-dej committed:
[dcl2] Move library code to `lib` subfolder

adam-dej committed:
[dcl2] Add (untested) posix threads helper

adam-dej committed:
[dcl2] Move unit tests to `tests/unit`

adam-dej committed:
Restructure makefile test targets to allow for more test sets

adam-dej committed:
[dcl2] Rename test suite to `dcl2-unit`

adam-dej committed:
Use `tee` for unit test output

adam-dej committed:
Add dcl2 pthread integration test targets to Makefile

adam-dej committed:
[dcl2-pthreads] Add first tests and fix discovered problems

adam-dej committed:
[lp] Add leaky-pipe header and docs

adam-dej committed:
Minor docs fixed

adam-dej committed:
Define leaky-pipe tests and general test improvements

adam-dej committed:
[leaky-pipe] Initial implementation and first test

adam-dej committed:
[leaky-pipe] Add tests

adam-dej committed:
Add valgrind suppression for invalid free in musl on Alpine

adam-dej committed:
[dcl2] Implement connection intgr test and fix intgr issue

At least these tests are good for something

adam-dej committed:
Fix failed tests not returning nonzero

adam-dej committed:
[leaky-pipe] Add pipe cutoff function

adam-dej committed:
[dcl2] Fix race in integration tests

adam-dej committed:
[dcl2] Add msg transmission intg test

adam-dej committed:
[dcl2] Split integration tests to multiple files

adam-dej committed:
Flush stdout after each test

adam-dej committed:
[dcl2] Add transmission tests over faulty links

adam-dej committed:
[dcl2] Fix frame retransmit acks not being sent

adam-dej committed:
[leaky-pipe] make lp_cutoff callable multiple times

adam-dej committed:
[dcl2] Implement integration test threaded assert helpers

adam-dej committed:
[dcl2] Start refactoring multithreaded integration tests

adam-dej committed:
[dcl2] Finish refactoring integration tests

adam-dej committed:
[dcrcp] Update protocol specification. Use CBOR

lubo commented on dcl2/lib/inc/dcl2.h:
```diff
@@ -0,0 +1,276 @@
+ *         DC_E_NOMSG if no message is waiting or
+ *         DC_E_FAIL  if parameters are invalid
+ *
+ * @note   Since "dcGetReceivedMsg" and "dcGetReceivedMsg" are 2 different functions, one might
```
Did you mean *"dcGetReceivedMsg"* and *"dcGetReceivedMsgLen"*?

adam-dej replied:
Fixed

lubo commented on dcl2/lib/inc/dcl2.h:
```diff
@@ -0,0 +1,276 @@
+/**
+ * Get the received message.
+ *
+ * This function returns the received message, in any. When the message is copied this function
```
***if** any*?

adam-dej replied:
Fixed

lubo commented on dcl2/lib/src/dcl2.c:
```diff
@@ -0,0 +1,372 @@
+                        uint8_t resp_f[f_len];
+                        yahdlc_frame_data(&resp_ctrl, NULL, 0, resp_f, &f_len);
+                        deadcom->transmitBytes(resp_f, f_len);
+                    }
```
Why did you put this block into braces?

adam-dej replied:
`resp_f` is an identiﬁer having a variably modiﬁed type, and the following `case` statement on line 358 would jump to its scope, which is prohibited by the C standard. Those braces create an explicit scope for `resp_f`.

lubo commented on docs/dcl2/protocol.md:
```diff
@@ -0,0 +1,270 @@
+Goal of this protocol is to provide a reliable datagram communication between
+exactly two devices connected using a point-to-point link.
+
+This protocol was heavily inspired by [HDLC (High-Level Data Link Control)](https://en.wikipedia.org/wiki/High-Level_Data_Link_Control), but simplified for our
```
How about using defining a block-level hyperlink target?

adam-dej replied:
I'm sorry, I'm not sure what you mean by a block-level hyperlink target (in markdown).

lubo replied:
Oh, I didn't noticed this is markdown. My bad, ignore my comment.

lubo commented on test/dcl2-integration/common.c:
```diff
@@ -0,0 +1,321 @@
+    rx_set_t *rp = (rx_set_t*) p;
+    uint8_t b[1];
+    while (lp_receive(rp->rx_pipe, b, 1)) {
+        // printf("Station %c received %02x\n", rp->station_char, b[0]);
```
Nobody likes code that is commented out without a obvious reason (at least it's not obvious to me).

adam-dej replied:
I've added a comment explaining what it may be useful for.

lubo reviewed (commented):
If you did create a separate repository for your yahdlc fork, it'd be easier to see your changes.

adam-dej committed:
Minor fixes suggested in a code review

adam-dej committed:
[dcrcp] Add implementation header

adam-dej committed:
[dcrcp] Stub implementation and add stub tests

adam-dej committed:
[dcrcp] Implement handling of SysQueryRequest CRPM

adam-dej committed:
[dcrcp] Update protocol CDDL and implement handling of other CRPMs

adam-dej committed:
[dcrcp] Switch cn-cbor branch to "memory-model"

This branch contains experimental changes to allocation context which
allow the use of allocators capable allocating only constant-size
objects.

adam-dej committed:
Start compiling tests with `-Wall` and `-Wextra`

adam-dej committed:
[dcl2] Fix types to have the correct semantic meaning

adam-dej committed:
[dcrcp] Use -Wall -Wextra for testing

adam-dej committed:
[dcrcp] Add const qualifiers to appropriate params

adam-dej committed:
Add python-related files to .gitignore

adam-dej committed:
[dcl2] Add void* context param to transmitBytes

This allows object-oriented approach to implementation of such function.

adam-dej committed:
[dcl2] [py3] Add stub python binding

adam-dej committed:
[dcl2] [py3] Implement python bindings for dcl2

adam-dej committed:
[dcl2] API BREAKING CHANGE make ext methods capable of failing

adam-dej committed:
[dcl2] [py3] Implement high-level pythonic wrapper for Deadcom

adam-dej committed:
[dcl2] [py3] Add documentation

adam-dej committed:
[dcl2] Limit max frame size yahdlc can parse

adam-dej committed:
[dcl2] Handle spurious wakeups in pthread helper lib

adam-dej commented:
This PR is no longer WIP, please perform a second round of review.

adam-dej committed:
[dcl2] [py3] Add CI config for extension building

adam-dej committed:
[dcl2] fix dcGetReceivedMsg not returning DC_NOT_CONNECTED

adam-dej committed:
[dcl2] [py3] Fix processing incorrect retcode from dcGetReceivedMsg

adam-dej committed:
[dcrcp] Add Heartbeat CRPM

adam-dej committed:
[dcrcp] Zero CRPM struct in decoder before decoding

adam-dej committed:
Merge pull request #3 from fmfi-svt-deadlock/feature-dcl2-python-bindings

Python bindings for dcl2

adam-dej committed:
[dcrcp] Minor fixes suggested in a review

adam-dej committed:
Merge pull request #2 from fmfi-svt-deadlock/feature-dcrcp

Specify and implement DCRCP
